components/LockScreen.tsx

```diff
@@ -34,18 +77,41 @@ export function LockScreen() {
   }, [lockMode, biometricAttempted, handleBiometricAuth]);
 
   const handlePinComplete = async (pin: string) => {
-    const isValid = await verifyPin(pin);
+    if (isPinDisabled || isLockedOut) return;
+
+    const result = await verifyPin(pin);
 
-    if (!isValid) {
-      setErrorMessage(t('lockScreen.incorrectPin'));
-      setTimeout(() => setErrorMessage(''), 3000);
+    if (result.isLockedOut && result.lockoutUntil) {
+      setIsLockedOut(true);
+      setLockoutUntil(result.lockoutUntil);
+      setErrorMessage(t('lockScreen.tooManyAttempts'));
+    } else if (!result.success) {
+      const attemptsMsg = result.remainingAttempts
+        ? t('lockScreen.incorrectPinWithAttempts', {
+            attempts: result.remainingAttempts,
+          })
+        : t('lockScreen.incorrectPin');
+      
+      setErrorMessage(attemptsMsg);
+
+      // Apply progressive delay
+      if (result.delayMs && result.delayMs > 0) {
+        setIsPinDisabled(true);
+        setTimeout(() => {
+          setIsPinDisabled(false);
+          setErrorMessage('');
+        }, result.delayMs);
+      } else {
+        setTimeout(() => setErrorMessage(''), 3000);
```

Comment:
Not sure about these uses of `setTimeout` without taking their result to cancel them when needed. Isn't there a cleaner or more reliable approach?